renku/cli/_providers/__init__.py

```diff
@@ -19,35 +19,56 @@
 from urllib.parse import urlparse
 
 from renku.cli._providers.zenodo import ZenodoProvider
+from renku.cli._providers.dataverse import DataverseProvider
 from renku.utils.doi import is_doi
 
 
 class ProviderFactory:
     """Create a provider type from URI."""
 
-    PROVIDERS = {'zenodo': ZenodoProvider}
+    PROVIDERS = {'dataverse': DataverseProvider, 'zenodo': ZenodoProvider}
 
     @staticmethod
     def from_uri(uri):
         """Get provider type based on uri."""
         is_doi_ = is_doi(uri)
-        if is_doi_ is False:
+        if is_doi_ is None:
             url = urlparse(uri)
             if bool(url.scheme and url.netloc and url.params == '') is False:
                 return None, 'Cannot parse URL.'
 
         provider = None
-        if 'zenodo' in uri:
-            provider = ZenodoProvider(is_doi=is_doi_)
+        warning = ''
+
+        for _, potential_provider in ProviderFactory.PROVIDERS.items():
+            try:
+                if potential_provider.supports(uri):
+                    provider = potential_provider
+                    break
+            except Exception as e:
```

Comment:
I think we should generally avoid catching generic `Exceptions`, but if we must then let's catch both (Exception, BaseException)

Reply:
The idea is that providers should handle expected exceptions themselves, so if there is any exception here, something went quite wrong. It isn't really necessary for the 2 providers we have now, but the more providers we have, the more difficult it'll be to account for all the possible interplay between them (Provider X might error when tested with an URI meant for Provider Y).

But it is relatively safe to assume that a provider will handle URIs (and exceptions) it was meant for, so if the provider is supposed to handle an URI, it should not throw some random exception for it.

In this case we show a warning, assume the provider does not support the URI, fall through to the next provider(s) and hopefully the URI can get handled successfully.

If no provider supports the URL, the warning is elevated to an exception.

This way, we should gracefully handle exceptions due to unforeseen edge-cases between providers without swallowing exceptions when we don't want to.

Good point on the `BaseException`.